_posts/2022-12-16-aave-rebasing.md

```diff
@@ -0,0 +1,332 @@
+be?
+
+If you were only looking at our deposits you might think we
+should have equal voting weight -- since we both deposited 100 UNI.
+But that's wrong.
+Ben should have _more_ voting weight than me because of all of the interest he
+earned over the last year.
+
+Ben's balance at the time of proposal was 150 aUNI.
+This would have entitled him to have withdrawn (and voted with) 150 UNI at that time.
+I, by contrast, could have only withdrawn and voted with the 100 UNI I had just deposited.
+
+So, it seems what we need to look at are _rebased balances_, not deposits.
+The rebased balances would have led us to the correct conclusion: Ben's voting preference should have 50% more voting weight than mine.
+If the aUNI contract holds 250 UNI and Ben is entitled to 150 of that, Ben
+should be able to determine 60% (150/250) of the votes aUNI casts, and I the
+remaining 40% (100/250).
+
+Thus it seems the dream scenario requires us to be able to
+calculate the rebased aToken balance for _any given user_ at _any given block_.
+
+In case it isn't obvious, this is very challenging.
+
+As already mentioned, aToken interest rates are constantly changing
+based on supply and demand within the protocol.
+If supply goes up, interest goes down.
+If demand goes up, interest goes up.
+
+And each time the interest rate changes, the interest accrued since the last update
+is added to the liquidity index (as shown above).
+This means that the aToken liquidity index is constantly changing too.
+
+Recall that the aToken's rebased balance is just the balance on disk multiplied by the liquidity index.
+
+If both values (balance-on-disk and liquidity index) are constantly changing,
+how can we hope to reliably compute a value based on them?
+
+Short of checkpointing both values _each time they change_ -- which would be
+unfeasibly expensive to write and later search for a popular asset with a lot of
+usage on Aave -- there seems to be no simple way to accurately compute an
+arbitrary user's aToken rebased balance at a block in the past.
+
+Fortunately, we don't have to.
+
+What we realized is that if we simply checkpoint each user's _scaled down balance_
+(i.e. their balance in storage) we can get all of the information we need to precisely
+calculate voting proportions.
+
+This is because the rebased balance is `scaled-down-balance x liquidity index`.
+Since every aToken holder's balance is being scaled up by the same liquidity
+index, the liquidity index can be ignored when computing ratios.
+For example, if the rebased balance for userA is `A x I` and the rebased balance for user B is
+`B x I` then the ratio of userA's balance to userB's is just `A / B` -- the index
+`I` cancels out.
+
+But if we only compare the stored values, wouldn't Ben and I end up with
+the same voting weight since we deposited the same amount?
+
+No.
+
+This is because Aave divides incoming balances by the current liquidity index before
+storing them.
```

Comment:
Feels like a lot of this section repeats info from the earlier rebasing section. IMO these two sections can be merged since voting weight is the same as user balances

Reply:
Yeah, I know what you mean. Let me see what I can do

Reply:
Tried to do this in 8468901